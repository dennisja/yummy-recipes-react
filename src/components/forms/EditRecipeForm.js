import React, {Component} from 'react';
import ReactDOM from 'react-dom';
import $ from 'jquery';
import {Redirect} from 'react-router-dom';

import CategoryRequest from '../../helpers/Categories';
import RecipesRequest from '../../helpers/Recipes';
import PreLoader, {displayError} from '../Utilities';

const CategoryOptions = (props) => {
    const {categories} = props;
    const options = categories.map(category => (
        <option value={category.id} key={category.id}>
            {category.name}
        </option>
    ))
    return options;
}

class EditRecipeForm extends Component {
    static initialState = {
        name: "",
        steps: "",
        ingredients: "",
        category: "",
        categories: [],
        loading: true,
        redirectToRecipes: false
    }

    state = EditRecipeForm.initialState;

    componentDidMount() {
        const {
            name,
            steps,
            ingredients,
            category: {
                id: category
            }
        } = this.props;
        CategoryRequest
            .fetchUserCategories()
            .then(response => {
                this.setState({
                    category,
                    steps,
                    ingredients,
                    name,
                    categories: response.data.recipe_cats,
                    loading: false
                })

                //initialize materialize select and overcome the select pain
                const el = ReactDOM.findDOMNode(this.refs.cat);
                $('select').material_select();
                $(el).on('change', this.handleInputChange)
            })
            .catch(error => {
                displayError(error);
            })
    }

    componentDidUpdate() {
        if (!this.state.loading) {
            const el = ReactDOM.findDOMNode(this.refs.cat);
            $('select').material_select();
            $(el).on('change', this.handleInputChange)
        }
    }

    handleRecipeSubmit = (event) => {
        event.preventDefault();
        const {
            categories,
            loading,
            ...recipeData
        } = this.state;

        RecipesRequest
            .editRecipe(this.props.id, recipeData)
            .then(response => {
                window
                    .Materialize
                    .toast(response.data.message, 4000);
                this.setState({redirectToRecipes: true})
            })
            .catch(error => {
                displayError(error);
            })
    }

    handleInputChange = (event) => {
        const {name, value} = event.target;
        this.setState({[name]: value})
    }

    render() {
        const {
            name,
            steps,
            ingredients,
            category,
            categories,
            loading,
            redirectToRecipes
        } = this.state;

        if (loading) {
            return (<PreLoader message="Please wait....."/>)
        }

        if (redirectToRecipes) {
            return <Redirect to="/recipes"/>
        }

        return (
            <form onSubmit={this.handleRecipeSubmit}>
                <div className="input-field">
                    <label htmlFor="name">Name</label>
                    <input
                        type="text"
                        name="name"
                        id="name"
                        value={name}
                        onChange={this.handleInputChange}/>
                </div>
                <div className="input-field">
                    <label htmlFor="ingredients">Ingredients</label>
                    <input
                        type="text"
                        name="ingredients"
                        id="ingredients"
                        value={ingredients}
                        onChange={this.handleInputChange}/>
                </div>
                {/* Edit textarea for steps */}
                <div className="input-field">
                    <textarea
                        value={steps}
                        onChange={this.handleInputChange}
                        name="steps"
                        id="steps"
                        className="materialize-textarea"/>
                    <label htmlFor="steps">Steps</label>
                </div>
                {/* Edit select field for category */}
                <div className="input-field">
                    <select
                        name="category"
                        ref="cat"
                        id="category"
                        value={category}
                        onChange={this.handleInputChange}>
                        <CategoryOptions categories={categories}/>
                    </select>
                    <label htmlFor="category">Category</label>
                </div>
                <div className="center-align">
                    <input type="submit" value="Edit Recipe" className="btn btn-small orange"/>
                </div>
            </form>
        );
    }
}

export default EditRecipeForm;