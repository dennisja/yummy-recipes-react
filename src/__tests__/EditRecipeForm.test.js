import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import $ from 'jquery';

import EditRecipeForm, { CategoryOptions } from '../components/forms/EditRecipeForm';

const categories = [
    {
        name: 'b',
        id: 1,
    }, {
        name: 'a',
        id: 2,
    }, {
        name: 'a',
        id: 3,
    },
];

describe('Test EditRecipeForm component', () => {
    const recipe = {
        name: 'h',
        steps: 'q',
        ingredients: 'i',
        category: {
            id: 12,
        },
        id: 1,
    };

    $('select').material_select = jest.fn();

    const formWrapper = shallow(<EditRecipeForm {...recipe} />);

    it('renders properly', () => {
        expect(shallowToJson(formWrapper)).toMatchSnapshot();
    });

    it('renders a PreLoader', () => {
        expect(formWrapper.find('PreLoader')).toHaveLength(1);
    });

    it('renders a form element', () => {
        formWrapper.setState({ loading: false, categories });
        formWrapper.update();
        expect(formWrapper.find('form').length).toBe(1);
    });

    it('renders the required form inputs', () => {
        formWrapper.setState({ loading: false, categories });
        formWrapper.update();
        expect(formWrapper.find('#name').length).toBe(1);
        expect(formWrapper.find('#ingredients').length).toBe(1);
        expect(formWrapper.find('#steps').length).toBe(1);
        expect(formWrapper.find('#category').length).toBe(1);
        expect(formWrapper.find('#submit_recipe').length).toBe(1);
    });
});

describe('Test Category options form', () => {
    const optionsWrapper = shallow(<CategoryOptions categories={categories} />);
    it('renders properly', () => {
        expect(shallowToJson(optionsWrapper)).toMatchSnapshot();
    });

    // commented because enzyme doesn't support array returning elements and fragments
    // it('renders the correct number of options', () => {
    //     expect(optionsWrapper.find('option')).toHaveLength(categories.length);
    // });

    // it('doesnot modify its props', () => {
    //     expect(optionsWrapper.instance().props.categories).toEqual(categories);
    // });
});
