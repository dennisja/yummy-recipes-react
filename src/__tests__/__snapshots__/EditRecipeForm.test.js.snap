// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Category options form renders properly 1`] = `
Array [
  <option
    key="1"
    value={1}
/>,
  <option
    key="2"
    value={2}
/>,
  <option
    key="3"
    value={3}
/>,
]
`;

exports[`Test EditRecipeForm component renders properly 1`] = `
<PreLoader
  message="Please wait....."
/>
`;
